### PR TITLE
Add emulation smoke/regression/extended tests.

### PR DESCRIPTION
**Pull request by geotseng-amd**
# Emulation Smoke/Regression/Extended Tests
---
## Usage -- rtest.py
Add emulation tests.
```
-   smoke test: rtest.py --emulation smoke --install_dir ${PATH_TO_hipblaslt-test}
-   regression test: rtest.py --emulation regression --install_dir ${PATH_TO_hipblaslt-test}
-   extended test: : rtest.py --emulation extended--install_dir ${PATH_TO_hipblaslt-test}
  ```
## Smoke-test yaml

Cover functionality and APIs for each Library to see that things ‘work’

- Should cover as many data types and APIs/features as possible, instead of covering every size permutation and edge cases that is normally run in CI
- Should cover any new data types that might be present on the new ASIC
- Should not cover every size permutation as the emulator runs extremely slowly, between 100x to 10,000x the speed of silicon.

## Runtime
- Use the following command to run the smoke/regression/extended tests.
``` python3 rtest.py --emulation smoke --install_dir build/release/clients/staging```
- The total filter must be finished in less than 1 minute using MI300 silicon (GPU time).
```
[----------] Global test environment tear-down
[==========] 2760 tests from 1 test suite ran. (32580 ms total)
[  PASSED  ] 2760 tests.
hipBLASLt version: 1300

command line: build/release/clients/staging/hipblaslt-test --gtest_filter=*smoke* --gtest_output=xml 
```
## Tickets:
1. hipBLASLt Emulation Smoke Tests: https://ontrack-internal.amd.com/browse/SWDEV-481982
2. hipBLASLt Emulation Regression Tests: https://ontrack-internal.amd.com/browse/SWDEV-484899
3. hipBLASLt Emulation Extended Tests: https://ontrack-internal.amd.com/browse/SWDEV-485109

### CODE REVIEW

**Comment by geotseng-amd** [comment]
All the local CI tests are passed.

[0;32m[----------] [mGlobal test environment tear-down
[0;32m[==========] [m54439 tests from 13 test suites ran. (1036697 ms total)
[0;32m[  PASSED  ] [m54439 tests.
hipBLASLt version: 1300

command line: ./build/release/clients/staging/hipblaslt-test --gtest_output=xml --gtest_color=yes --gtest_repeat=1

File: clients/gtest/smoke_gtest.yaml [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+  scaleC: [0,1]
+  scaleD: [0,1]
+  use_ext: [1]
+  use_ext_setproblem: [1]
+  bias_vector: 1
+  bias_type: [f32_r]
+  unit_check: 1
+  gpu_arch: '94[0-2]'
+
+- name: matmul_gemm_amaxAB_mix_precisions_fp8_smoke
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: remove mix precision

File: clients/gtest/smoke_gtest.yaml [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+  - &alpha_beta_range_small
+    - { alpha: 2, alphai: 2, beta: -1.0, betai: 2.0 }
+
+  - &transA_transB_range
+    - { transA: N, transB: N }
+    - { transA: N, transB: T }
+    - { transA: T, transB: N }
+    - { transA: T, transB: T }
+
+  - &deepbench_alpha_beta_range
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: remove deepbench related configurations

File: clients/gtest/smoke_gtest.yaml [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+    - { alpha: 1, beta: 0 }
+    - { alpha: 1, beta: 1 }
+
+  - &deepbench_transA_transB_range
+    - { transA: N, transB: N }
+    - { transA: N, transB: T }
+    - { transA: T, transB: N }
+
+  - &ldd_transA_transB_range
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: same with transA_transB_range, remove it

File: clients/gtest/smoke_gtest.yaml [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+- name: matmul_small_smoke
+  category: smoke
+  function:
+    matmul: *real_precisions
+  matrix_size: *smoke_matrix_size_range
+  transA_transB: *transA_transB_range
+  alpha_beta: *alpha_beta_range
+  c_equal_d: [false, true]
+
+- name: matmul_conj_small_smoke
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: conjungate is  tested is API level test. remove it

File: clients/gtest/smoke_gtest.yaml [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+  scaleB: [0, 1]
+  scaleC: [1] # will use random value from 0.0 to 1.0
+  scaleD: [1] # will use random value from 0.0 to 1.0
+  bias_vector: [0, 1]
+  bias_type: f16_r
+  unit_check: 0
+  norm_check: 1
+  gpu_arch: '94[0-2]'
+
+- name: matmul_real_1b_dst_1b_amaxD_smoke
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: remove amaxD test

File: clients/gtest/smoke_gtest.yaml [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+    matmul: *hpa_half_precisions_compute_fp16
+  matrix_size:
+    - { M:  128,  N:  128,  K:  128  }
+  transA: N
+  transB: N
+  norm_check: 1
+  unit_check: 0
+  gpu_arch: '110?'
+
+- name: matmul_input_fp16_computeIn_bf16_smoke
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: remove this, this is special case

File: clients/gtest/smoke_gtest.yaml [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+    matmul: *real_precisions_2b
+  matrix_size: *smoke_matrix_size_range
+  transA_transB: *transA_transB_range
+  alpha: 1
+  beta: [ 0.0, 2.0 ]
+  bias_vector: 1
+  bias_type: [default, f32_r]
+  unit_check: 1
+
+- name: matmul_fallback_compute_fp16_smoke
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: this is API level test, remove it

File: clients/gtest/smoke_gtest.yaml [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+  transA_transB: *transA_transB_range
+  alpha: 1
+  beta: [ 0, 2 ]
+  bias_vector: [0, 1]
+  scaleAlpha_vector: [0, 1]
+  bias_type: [f16_r, f32_r]
+  unit_check: 1
+  gpu_arch: '90a'
+
+- name: matmul_input_fp16_computeIn_fp8_smoke
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: remove this

File: clients/gtest/smoke_gtest.yaml [hunk 9]
```diff
@@ -0,0 +1,574 @@
...
+  alpha: 1
+  beta: [ 0.0, 2.0 ]
+  scaleA: [ 0, 1]
+  scaleB: [ 0, 1]
+  bias_vector: [0, 1]
+  bias_type: bf16_r
+  unit_check: 1
+  gpu_arch: '94[0-2]'
+
+- name: matmul_f8_dst_bf16_scale_ab_vec_smoke
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: remove this

File: clients/gtest/smoke_gtest.yaml (hunk 1 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: clients/gtest/smoke_gtest.yaml (hunk 2 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: clients/gtest/smoke_gtest.yaml (hunk 3 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: clients/gtest/smoke_gtest.yaml (hunk 4 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: clients/gtest/smoke_gtest.yaml (hunk 5 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: clients/gtest/smoke_gtest.yaml (hunk 6 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: DOne

File: clients/gtest/smoke_gtest.yaml (hunk 7 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: clients/gtest/smoke_gtest.yaml (hunk 8 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: clients/gtest/smoke_gtest.yaml (hunk 9 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: clients/gtest/smoke_gtest.yaml [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+  alpha: 1
+  beta: [ 0, 2 ]
+  use_ext: [1]
+  use_ext_setproblem: [1]
+  bias_vector: 1
+  bias_type: [f32_r]
+  unit_check: 1
+  gpu_arch: '94[0-2]'
+
+- name: matmul_gemm_mix_precisions_fp8_smoke
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: remove mix precision

File: clients/gtest/smoke_gtest.yaml [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+    - { M:  128,  N:  128,  K:  128  }
+  transA_transB: *transA_transB_range
+  alpha: 1
+  beta: [ 0, 2 ]
+  scaleAlpha_vector: [0, 1]
+  activation_type: relu
+  unit_check: 1
+  gpu_arch: '1[1-2]\d{2}'
+
+- name: matmul_gemm_mix_precisions_smoke
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: remove mix precision

File: clients/gtest/smoke_gtest.yaml [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+    - { transA: N, transB: T }
+    - { transA: T, transB: N }
+    - { transA: T, transB: T }
+
+Tests:
+- name: matmul_smoke
+  category: smoke
+  function:
+    matmul: *real_precisions
+  matrix_size: *medium_matrix_size_range
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: why do you still use medium size for smoke test?

File: clients/gtest/smoke_gtest.yaml (hunk 11 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: clients/gtest/smoke_gtest.yaml (hunk 10 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: clients/gtest/smoke_gtest.yaml (hunk 12 above)
Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done, modified.

File: clients/gtest/smoke_gtest.yaml [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+    matmul: *real_precisions_2b
+  matrix_size: *smoke_matrix_size_range
+  transA_transB: *transA_transB_range
+  alpha: 1
+  beta: [ 0.0, 2.0 ]
+  bias_vector: 1
+  bias_type: [default, f32_r]
+  unit_check: 1
+
+- name: matmul_fallback_compute_fp16_gfx11_smoke
```
Comment (jichangjichang) [inline clients/gtest/smoke_gtest.yaml]: remove fallback test

Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: Done

File: rtest.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+def run_test(args):
+
+    cwd = os.curdir
+
+    if args.emulation == "smoke":
+        run_cmd(args, "--gtest_filter=*smoke*")
+    elif args.emulation == "regression":
+        run_cmd(args, "--gtest_filter=*quick*")
+    elif args.emulation == "extended":
+         print("There is no test for extended test cases.")
```
Comment (jichangjichang) [inline rtest.py]: We can use precheck_in and nightly for extended tests

> --gtest_filter=\*pre_checkin*:\*nightly*

Comment (geotseng-amd) [reply rtest.py]: sure

File: clients/gtest/smoke_gtest.yaml (hunk 10 above)
Comment (jichangjichang) [reply clients/gtest/smoke_gtest.yaml]: it is still there

Comment (geotseng-amd) [reply clients/gtest/smoke_gtest.yaml]: I will commit after all the modifications are done. Thank you.

File: rtest.py (hunk 14 above)
Comment (geotseng-amd) [reply rtest.py]: Done

**Comment by geotseng-amd** [comment]
[0;32m[----------] [mGlobal test environment tear-down
[0;32m[==========] [m53102 tests from 13 test suites ran. (1997701 ms total)
[0;32m[  PASSED  ] [m53102 tests.
hipBLASLt version: 1300

command line: ./build/release/clients/staging/hipblaslt-test --gtest_output=xml --gtest_color=yes --gtest_repeat=1

**Review by jichangjichang — APPROVED** [review]

**Comment by cliffxzx** [comment]
gfx1100
```
[----------] Global test environment tear-down
[==========] 14563 tests from 13 test suites ran. (743898 ms total)
[  PASSED  ] 14563 tests.
hipBLASLt version: 1300

command line: build/742aea/debug/clients/staging/hipblaslt-test 
```

gfx1201
image

**Review by Serge45 — APPROVED** [review]
